### ISSUE TITLE
Inconsistent generation with flash attention disabled

### ISSUE THREAD

**Issue by shimizust** [huggingface/text-generation-inference#1262]
### System Info

```sh
2023-11-15T00:37:21.574339Z  INFO text_generation_launcher: Runtime environment:
Target: x86_64-unknown-linux-gnu
Cargo version: 1.70.0
Commit sha: 96a982ad8fc232479384476b1596a880697cc1d0
Docker label: sha-96a982a
nvidia-smi:
Wed Nov 15 00:37:21 2023       
   +---------------------------------------------------------------------------------------+
   | NVIDIA-SMI 530.30.02              Driver Version: 530.30.02    CUDA Version: 12.1     |
   |-----------------------------------------+----------------------+----------------------+
   | GPU  Name                  Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
   | Fan  Temp  Perf            Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
   |                                         |                      |               MIG M. |
   |=========================================+======================+======================|
   |   0  NVIDIA A100-SXM4-80GB           On | 00000000:8A:00.0 Off |                    0 |
   | N/A   27C    P0               73W / 400W|  79571MiB / 81920MiB |      0%      Default |
   |                                         |                      |             Disabled |
   +-----------------------------------------+----------------------+----------------------+
                                                                                            
   +---------------------------------------------------------------------------------------+
   | Processes:                                                                            |
   |  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
   |        ID   ID                                                             Usage      |
   |=======================================================================================|
   +---------------------------------------------------------------------------------------+
2023-11-15T00:37:21.574394Z  INFO text_generation_launcher: Args { model_id: "bigscience/bloom-560m", revision: None, validation_workers: 2, sharded: None, num_shard: None, quantize: None, dtype: None, trust_remote_code: false, max_concurrent_requests: 128, max_best_of: 2, max_stop_sequences: 4, max_top_n_tokens: 5, max_input_length: 1024, max_total_tokens: 2048, waiting_served_ratio: 1.2, max_batch_prefill_tokens: 4096, max_batch_total_tokens: None, max_waiting_tokens: 20, hostname: "tgi-test", port: 80, shard_uds_path: "/tmp/text-generation-server", master_addr: "localhost", master_port: 29500, huggingface_hub_cache: Some("/data"), weights_cache_override: None, disable_custom_kernels: false, cuda_memory_fraction: 1.0, rope_scaling: None, rope_factor: None, json_output: false, otlp_endpoint: None, cors_allow_origin: [], watermark_gamma: None, watermark_delta: None, ngrok: false, ngrok_authtoken: None, ngrok_edge: None, env: true }
2023-11-15T00:37:21.575003Z  INFO download: text_generation_launcher: Starting download process.
```

Running on A100 on k8s cluster. 

### Information

- [X] Docker
- [ ] The CLI directly

### Tasks

- [X] An officially supported command
- [ ] My own modifications

### Reproduction

1. Run this command in the TGI docker container: `text-generation-launcher --model-id "tiiuae/falcon-7b" --port 8080 --trust-remote-code` with env variable "USE_FLASH_ATTENTION" either set to "true" or "false".
2. Port-forward to tgi server
3. Open up Swagger UI to test: `localhost:8080`

Try the following curl request (with seed value specified):
```
curl -X 'POST' \
  'http://localhost:8080/generate' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "inputs": "My name is Olivier and I",
  "parameters": {
    "best_of": 1,
    "decoder_input_details": true,
    "details": true,
    "do_sample": true,
    "max_new_tokens": 20,
    "repetition_penalty": 1.03,
    "return_full_text": false,
    "seed": 123456,
    "stop": [
      "photographer"
    ],
    "temperature": 0.5,
    "top_k": 10,
    "top_n_tokens": 5,
    "top_p": 0.95,
    "truncate": null,
    "typical_p": 0.95,
    "watermark": true
  }
}'
```

In the case where flash attention is enabled (default), I get reasonable response:
```
{
  "generated_text": "’m a freelance graphic designer, illustrator and artist.\nMy work consists in designing and illustrating for",
  "details": {
    "finish_reason": "length",
    "generated_tokens": 20,
    "seed": 123456,
   ...
}
```

In the case where flash attention is disabled, I get nonsense, repeated output:
```
{
  "generated_text": "’.\n (\n (\n ( son 6.\n\n 6,",
  "details": {
    "finish_reason": "length",
    "generated_tokens": 20,
    "seed": 123456,
    ...
}
```

### Expected behavior

I would expect the same output in both cases (or at least similarly good output). My understanding is that flash attention improves performance but shouldn't affect the generation results. When flash attention is disabled, should that effectively be the same as using AutoModel?

**Comment by OlivierDehaene** [comment]
Falcon with AutoModel might be bugged since the repo were modified to match the implementation in Transformers. AutoModel is best effort and we do not do regression testing on them. 
Can you check if `--revision 2f5c3cd4eace6be6c0f12981f377fb35e5bf6ee5` works with AutoModel?

**Comment by shimizust** [comment]
Thanks @OlivierDehaene . I realized we are using revision `da8d49a4c7dde3bfc39461e6f2cf7433e2fa44c2` of falcon-7b, which seems to be effectively the same as `2f5c3cd4eace6be6c0f12981f377fb35e5bf6ee5`, which was before the repo was modified to use the new Falcon model implementations in transformers. Let me try the latest version.

The actual reason behind this issue was when we load this same version of falcon-7b using transformers  (v4.33.1) AutoModelForCausalLM directly for batch inference, it produces decently good output but still different (and slightly worse) than TGI. I was trying to isolate the issue by turning off flash attention in TGI to force it to use the same AutoModel, but the generated output is much worse.

To summarize the quality of generation:
`TGI (flash attn enabled) > transformers AutoModel >> TGI (flash attn disabled)`

Is TGI doing something different than just passing off the model instantiation to the transformers library which could be causing this discrepancy?

**Comment by haodeqi** [comment]
I am able to observe a difference with `tiiuae/falcon-7b` with & w/o flash attention. When flash attention is not enabled, half of output from selfattention is nan value in my particular test sample.
